File: internal/sources/vscode/empty_window.go
package vscode

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/pcr-developers/cli/internal/display"
	"github.com/pcr-developers/cli/internal/sources/shared"
	"github.com/pcr-developers/cli/internal/store"
	"github.com/pcr-developers/cli/internal/supabase"
	"github.com/pcr-developers/cli/internal/versions"
)

// ─── V3 JSON format (legacy) ─────────────────────────────────────────────────

type emptyWindowSession struct {
	Version       int                `json:"version"`
	SessionID     string             `json:"sessionId"`
	CreationDate  int64              `json:"creationDate"` // unix ms
	CustomTitle   string             `json:"customTitle"`
	Requests      []emptyWindowReq   `json:"requests"`
}

type emptyWindowReq struct {
	Message    emptyWindowMsg      `json:"message"`
	Response   []emptyWindowResp   `json:"response"`
	Result     *emptyWindowResult  `json:"result"`
	Agent      *emptyWindowAgent   `json:"agent"`
	IsCanceled bool                `json:"isCanceled"`
}

type emptyWindowMsg struct {
	Parts []emptyWindowPart `json:"parts"`
}

type emptyWindowPart struct {
	Text string `json:"text"`
}

type emptyWindowResp struct {
	Value string `json:"value"`
}

type emptyWindowResult struct {
	Timings  *emptyWindowTimings  `json:"timings"`
	Metadata json.RawMessage      `json:"metadata"`
}

type emptyWindowTimings struct {
	FirstProgress int64 `json:"firstProgress"`
	TotalElapsed  int64 `json:"totalElapsed"`
}

type emptyWindowAgent struct {
	ID string `json:"id"`
}

// ─── JSONL mutation log format ───────────────────────────────────────────────

type mutationEntry struct {
	Kind  int             `json:"kind"`
	Key   []string        `json:"k"`
	Value json.RawMessage `json:"v"`
}

type emptyWindowSnapshot struct {
	Version      int              `json:"version"`
	SessionID    string           `json:"sessionId"`
	CreationDate int64            `json:"creationDate"`
	Requests     []emptyWindowReq `json:"requests"`
	InputState   *inputState      `json:"inputState"`
}

type inputState struct {
	Mode          *modeInfo          `json:"mode"`
	SelectedModel *selectedModelInfo `json:"selectedModel"`
	PermLevel     string             `json:"permissionLevel"`
}

type modeInfo struct {
	ID   string `json:"id"`
	Kind string `json:"kind"`
}

type selectedModelInfo struct {
	Identifier string         `json:"identifier"`
	Metadata   *modelMetadata `json:"metadata"`
}

type modelMetadata struct {
	ID     string `json:"id"`
	Name   string `json:"name"`
	Vendor string `json:"vendor"`
}

// ProcessEmptyWindowSessions scans the emptyWindowChatSessions directory for
// sessions without a workspace and saves new exchanges as drafts.
func ProcessEmptyWindowSessions(userID string, state *shared.FileState, dedup *shared.Deduplicator) {
	globalBase := GlobalStorageBase()
	if globalBase == "" {
		return
	}
	dir := filepath.Join(globalBase, "emptyWindowChatSessions")
	entries, err := os.ReadDir(dir)
	if err != nil {
		return
	}

	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		path := filepath.Join(dir, e.Name())
		name := e.Name()

		if strings.HasSuffix(name, ".json") {
			processJSONSession(path, userID, state, dedup)
		} else if strings.HasSuffix(name, ".jsonl") {
			processJSONLSession(path, userID, state, dedup)
		}
	}
}

// processJSONSession handles the legacy v3 JSON format.
func processJSONSession(filePath, userID string, state *shared.FileState, dedup *shared.Deduplicator) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return
	}

	// Use file size as state marker
	prevSize := state.Get(filePath)
	if len(data) <= prevSize {
		return
	}
	state.Set(filePath, len(data))

	var session emptyWindowSession
	if err := json.Unmarshal(data, &session); err != nil {
		return
	}
	if session.Version < 3 || len(session.Requests) == 0 {
		return
	}

	saveEmptyWindowExchanges(session.SessionID, session.CreationDate, session.Requests, "", "", userID, dedup)
}

// processJSONLSession handles the kind-based mutation log format.
func processJSONLSession(filePath, userID string, state *shared.FileState, dedup *shared.Deduplicator) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return
	}

	prevSize := state.Get(filePath)
	if len(data) <= prevSize {
		return
	}
	state.Set(filePath, len(data))

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	// Replay ALL mutations into a generic JSON tree so responses,
	// new requests, and other property updates are not lost.
	var tree any

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var entry mutationEntry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			continue
		}

		switch entry.Kind {
		case 0:
			// Full snapshot — replace entire tree
			var val any
			if err := json.Unmarshal(entry.Value, &val); err == nil {
				tree = val
			}
		case 1:
			// Property mutation — set value at key path
			if tree != nil && len(entry.Key) > 0 {
				tree = applyMutation(tree, entry.Key, entry.Value)
			}
		}
	}

	if tree == nil {
		return
	}

	// Marshal reconstructed tree back to structured type
	treeJSON, err := json.Marshal(tree)
	if err != nil {
		return
	}

	var snapshot emptyWindowSnapshot
	if err := json.Unmarshal(treeJSON, &snapshot); err != nil {
		return
	}

	var modelName string
	if snapshot.InputState != nil && snapshot.InputState.SelectedModel != nil {
		if snapshot.InputState.SelectedModel.Metadata != nil {
			modelName = snapshot.InputState.SelectedModel.Metadata.Name
		} else {
			modelName = snapshot.InputState.SelectedModel.Identifier
		}
	}

	if snapshot.SessionID == "" || len(snapshot.Requests) == 0 {
		return
	}

	saveEmptyWindowExchanges(snapshot.SessionID, snapshot.CreationDate, snapshot.Requests, modelName, "", userID, dedup)
}

// applyMutation sets a value at the given key path in a generic JSON tree.
// Handles both map[string]any (objects) and []any (arrays).
func applyMutation(root any, keys []string, rawValue json.RawMessage) any {
	if len(keys) == 0 {
		var val any
		_ = json.Unmarshal(rawValue, &val)
		return val
	}

	key := keys[0]
	rest := keys[1:]

	// Try to interpret key as array index
	if idx, err := strconv.Atoi(key); err == nil && idx >= 0 {
		arr, ok := root.([]any)
		if !ok {
			arr = []any{}
		}
		for len(arr) <= idx {
			arr = append(arr, nil)
		}
		arr[idx] = applyMutation(arr[idx], rest, rawValue)
		return arr
	}

	// Object key
	m, ok := root.(map[string]any)
	if !ok {
		m = map[string]any{}
	}
	m[key] = applyMutation(m[key], rest, rawValue)
	return m
}

// saveEmptyWindowExchanges converts request/response pairs into drafts.
func saveEmptyWindowExchanges(sessionID string, creationDateMs int64, requests []emptyWindowReq, model, mode, userID string, dedup *shared.Deduplicator) {
	createdAt := time.UnixMilli(creationDateMs).UTC().Format(time.RFC3339)

	var newCount int
	for i, req := range requests {
		if req.IsCanceled {
			continue
		}
		promptText := extractEmptyWindowPrompt(req)
		if strings.TrimSpace(promptText) == "" {
			continue
		}
		responseText := extractEmptyWindowResponse(req)

		// Compute a per-exchange timestamp (offset by index if no better data)
		capturedAt := createdAt
		if creationDateMs > 0 {
			// Rough approximation: offset by request index × 30s
			t := time.UnixMilli(creationDateMs).Add(time.Duration(i) * 30 * time.Second)
			capturedAt = t.UTC().Format(time.RFC3339)
		}

		hash := supabase.PromptContentHashV2(sessionID, promptText, capturedAt)
		if dedup.IsDuplicate(sessionID, hash) {
			continue
		}
		if store.IsDraftSavedAt(sessionID, promptText, capturedAt) {
			dedup.Mark(sessionID, hash)
			continue
		}

		fileContext := map[string]any{
			"capture_schema": versions.CaptureSchemaVersion,
			"is_agentic":     req.Agent != nil && req.Agent.ID != "",
		}
		if req.Result != nil && req.Result.Timings != nil {
			fileContext["response_duration_ms"] = req.Result.Timings.TotalElapsed
			if req.Result.Timings.FirstProgress > 0 {
				fileContext["first_response_ms"] = req.Result.Timings.FirstProgress
			}
		}

		record := supabase.PromptRecord{
			ID:            supabase.PromptIDV2(sessionID, promptText, capturedAt),
			ContentHash:   hash,
			SessionID:     sessionID,
			PromptText:    promptText,
			ResponseText:  responseText,
			Model:         model,
			Source:        "vscode",
			CaptureMethod: "file-watcher",
			CapturedAt:    capturedAt,
			UserID:        userID,
			FileContext:   fileContext,
		}

		if err := store.SaveDraft(record, nil, ""); err != nil {
			display.PrintError("vscode", fmt.Sprintf("Failed to save empty-window draft: %s", err.Error()))
			continue
		}
		dedup.Mark(sessionID, hash)
		newCount++
	}

	if newCount > 0 {
		display.PrintDrafted(display.DraftDisplayOptions{
			ProjectName:   "(no workspace)",
			PromptText:    extractEmptyWindowPrompt(requests[len(requests)-1]),
			ExchangeCount: newCount,
		})
	}
}

func extractEmptyWindowPrompt(req emptyWindowReq) string {
	var parts []string
	for _, p := range req.Message.Parts {
		if p.Text != "" {
			parts = append(parts, p.Text)
		}
	}
	return strings.Join(parts, "\n")
}

func extractEmptyWindowResponse(req emptyWindowReq) string {
	var parts []string
	for _, r := range req.Response {
		if r.Value != "" {
			parts = append(parts, r.Value)
		}
	}
	return strings.Join(parts, "\n")
}

File: crates/pcr-core/src/sources/vscode/empty_window.rs
//! VS Code "empty window" Copilot chat session processor. Direct port of
//! `cli/internal/sources/vscode/empty_window.go`.
//!
//! Handles both the legacy v3 JSON format and the newer mutation-log JSONL
//! format (kind: 0 = full snapshot; kind: 1 = property mutation).

use serde::Deserialize;
use serde_json::{Map, Value};
use std::path::PathBuf;

use crate::display;
use crate::sources::shared::{Deduplicator, FileState};
use crate::sources::vscode::workspace::global_storage_base;
use crate::store::{self, is_draft_saved_at};
use crate::supabase::{self, PromptRecord};
use crate::versions;

// ─── Types matching the on-disk shape ────────────────────────────────────────

#[derive(Debug, Deserialize, Default)]
#[serde(default)]
struct EmptyWindowSession {
    version: i64,
    #[serde(rename = "sessionId")]
    session_id: String,
    #[serde(rename = "creationDate")]
    creation_date: i64,
    #[serde(rename = "customTitle")]
    #[allow(dead_code)]
    custom_title: String,
    requests: Vec<EmptyWindowReq>,
}

#[derive(Debug, Deserialize, Default, Clone)]
#[serde(default)]
struct EmptyWindowReq {
    message: EmptyWindowMsg,
    response: Vec<EmptyWindowResp>,
    result: Option<EmptyWindowResult>,
    agent: Option<EmptyWindowAgent>,
    #[serde(rename = "isCanceled")]
    is_canceled: bool,
}

#[derive(Debug, Deserialize, Default, Clone)]
#[serde(default)]
struct EmptyWindowMsg {
    parts: Vec<EmptyWindowPart>,
}

#[derive(Debug, Deserialize, Default, Clone)]
#[serde(default)]
struct EmptyWindowPart {
    text: String,
}

#[derive(Debug, Deserialize, Default, Clone)]
#[serde(default)]
struct EmptyWindowResp {
    value: String,
}

#[derive(Debug, Deserialize, Default, Clone)]
#[serde(default)]
struct EmptyWindowResult {
    timings: Option<EmptyWindowTimings>,
    #[allow(dead_code)]
    metadata: Value,
}

#[derive(Debug, Deserialize, Default, Clone)]
#[serde(default)]
struct EmptyWindowTimings {
    #[serde(rename = "firstProgress")]
    first_progress: i64,
    #[serde(rename = "totalElapsed")]
    total_elapsed: i64,
}

#[derive(Debug, Deserialize, Default, Clone)]
#[serde(default)]
struct EmptyWindowAgent {
    id: String,
}

#[derive(Debug, Deserialize, Default)]
#[serde(default)]
struct MutationEntry {
    kind: i64,
    k: Vec<String>,
    v: Value,
}

#[derive(Debug, Deserialize, Default)]
#[serde(default)]
struct EmptyWindowSnapshot {
    #[allow(dead_code)]
    version: i64,
    #[serde(rename = "sessionId")]
    session_id: String,
    #[serde(rename = "creationDate")]
    creation_date: i64,
    requests: Vec<EmptyWindowReq>,
    #[serde(rename = "inputState")]
    input_state: Option<InputState>,
}

#[derive(Debug, Deserialize, Default)]
#[serde(default)]
struct InputState {
    mode: Option<ModeInfo>,
    #[serde(rename = "selectedModel")]
    selected_model: Option<SelectedModelInfo>,
    #[serde(rename = "permissionLevel")]
    #[allow(dead_code)]
    perm_level: String,
}

#[derive(Debug, Deserialize, Default)]
#[serde(default)]
struct ModeInfo {
    #[allow(dead_code)]
    id: String,
    #[allow(dead_code)]
    kind: String,
}

#[derive(Debug, Deserialize, Default)]
#[serde(default)]
struct SelectedModelInfo {
    identifier: String,
    metadata: Option<ModelMetadata>,
}

#[derive(Debug, Deserialize, Default)]
#[serde(default)]
struct ModelMetadata {
    #[allow(dead_code)]
    id: String,
    name: String,
    #[allow(dead_code)]
    vendor: String,
}

// ─── Entry point ─────────────────────────────────────────────────────────────

pub fn process_empty_window_sessions(user_id: &str, state: &FileState, dedup: &Deduplicator) {
    let global_base = global_storage_base();
    let dir = global_base.join("emptyWindowChatSessions");
    let Ok(entries) = std::fs::read_dir(&dir) else {
        return;
    };
    for entry in entries.filter_map(|e| e.ok()) {
        let Ok(file_type) = entry.file_type() else {
            continue;
        };
        if file_type.is_dir() {
            continue;
        }
        let path = entry.path();
        let name = entry.file_name();
        let name_str = name.to_string_lossy();
        if name_str.ends_with(".json") {
            process_json_session(&path, user_id, state, dedup);
        } else if name_str.ends_with(".jsonl") {
            process_jsonl_session(&path, user_id, state, dedup);
        }
    }
}

fn process_json_session(
    file_path: &PathBuf,
    user_id: &str,
    state: &FileState,
    dedup: &Deduplicator,
) {
    let Ok(data) = std::fs::read(file_path) else {
        return;
    };
    let key = file_path.to_string_lossy().into_owned();
    let prev_size = state.get(&key);
    if (data.len() as i64) <= prev_size {
        return;
    }
    state.set(&key, data.len() as i64);

    let Ok(session) = serde_json::from_slice::<EmptyWindowSession>(&data) else {
        return;
    };
    if session.version < 3 || session.requests.is_empty() {
        return;
    }
    save_empty_window_exchanges(
        &session.session_id,
        session.creation_date,
        &session.requests,
        "",
        user_id,
        dedup,
    );
}

fn process_jsonl_session(
    file_path: &PathBuf,
    user_id: &str,
    state: &FileState,
    dedup: &Deduplicator,
) {
    let Ok(data) = std::fs::read(file_path) else {
        return;
    };
    let key = file_path.to_string_lossy().into_owned();
    let prev_size = state.get(&key);
    if (data.len() as i64) <= prev_size {
        return;
    }
    state.set(&key, data.len() as i64);

    let text = String::from_utf8_lossy(&data);
    let mut tree: Value = Value::Null;

    for line in text.trim().split('\n') {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let Ok(entry) = serde_json::from_str::<MutationEntry>(line) else {
            continue;
        };
        match entry.kind {
            0 => tree = entry.v,
            1 => {
                if !tree.is_null() && !entry.k.is_empty() {
                    tree = apply_mutation(tree, &entry.k, entry.v);
                }
            }
            _ => {}
        }
    }

    if tree.is_null() {
        return;
    }
    let Ok(snap) = serde_json::from_value::<EmptyWindowSnapshot>(tree) else {
        return;
    };

    let mut model_name = String::new();
    if let Some(state) = &snap.input_state {
        if let Some(sel) = &state.selected_model {
            model_name = match &sel.metadata {
                Some(m) => m.name.clone(),
                None => sel.identifier.clone(),
            };
        }
    }

    if snap.session_id.is_empty() || snap.requests.is_empty() {
        return;
    }
    save_empty_window_exchanges(
        &snap.session_id,
        snap.creation_date,
        &snap.requests,
        &model_name,
        user_id,
        dedup,
    );
}

fn apply_mutation(root: Value, keys: &[String], raw: Value) -> Value {
    if keys.is_empty() {
        return raw;
    }
    let key = &keys[0];
    let rest = &keys[1..];

    if let Ok(idx) = key.parse::<usize>() {
        let mut arr = match root {
            Value::Array(a) => a,
            _ => Vec::new(),
        };
        while arr.len() <= idx {
            arr.push(Value::Null);
        }
        let current = arr[idx].take();
        arr[idx] = apply_mutation(current, rest, raw);
        return Value::Array(arr);
    }

    let mut m = match root {
        Value::Object(m) => m,
        _ => Map::new(),
    };
    let current = m.remove(key).unwrap_or(Value::Null);
    let new_value = apply_mutation(current, rest, raw);
    m.insert(key.clone(), new_value);
    Value::Object(m)
}

fn save_empty_window_exchanges(
    session_id: &str,
    creation_date_ms: i64,
    requests: &[EmptyWindowReq],
    model: &str,
    user_id: &str,
    dedup: &Deduplicator,
) {
    let created_at = chrono::DateTime::<chrono::Utc>::from_timestamp_millis(creation_date_ms)
        .map(|dt| dt.to_rfc3339_opts(chrono::SecondsFormat::Secs, true))
        .unwrap_or_default();

    let mut new_count = 0usize;
    for (i, req) in requests.iter().enumerate() {
        if req.is_canceled {
            continue;
        }
        let prompt_text = extract_empty_window_prompt(req);
        if prompt_text.trim().is_empty() {
            continue;
        }
        let response_text = extract_empty_window_response(req);

        let captured_at = if creation_date_ms > 0 {
            let offset = creation_date_ms + (i as i64) * 30_000;
            chrono::DateTime::<chrono::Utc>::from_timestamp_millis(offset)
                .map(|dt| dt.to_rfc3339_opts(chrono::SecondsFormat::Secs, true))
                .unwrap_or_else(|| created_at.clone())
        } else {
            created_at.clone()
        };

        let hash = supabase::prompt_content_hash_v2(session_id, &prompt_text, &captured_at);
        if dedup.is_duplicate(session_id, &hash) {
            continue;
        }
        if is_draft_saved_at(session_id, &prompt_text, &captured_at) {
            dedup.mark(session_id, &hash);
            continue;
        }

        let mut file_context = serde_json::Map::new();
        file_context.insert(
            "capture_schema".into(),
            Value::Number(versions::CAPTURE_SCHEMA_VERSION.into()),
        );
        let is_agentic = req.agent.as_ref().is_some_and(|a| !a.id.is_empty());
        file_context.insert("is_agentic".into(), Value::Bool(is_agentic));
        if let Some(result) = &req.result {
            if let Some(timings) = &result.timings {
                file_context.insert(
                    "response_duration_ms".into(),
                    Value::Number(timings.total_elapsed.into()),
                );
                if timings.first_progress > 0 {
                    file_context.insert(
                        "first_response_ms".into(),
                        Value::Number(timings.first_progress.into()),
                    );
                }
            }
        }

        let record = PromptRecord {
            id: supabase::prompt_id_v2(session_id, &prompt_text, &captured_at),
            content_hash: hash.clone(),
            session_id: session_id.to_string(),
            prompt_text,
            response_text,
            model: model.to_string(),
            source: "vscode".into(),
            capture_method: "file-watcher".into(),
            captured_at,
            user_id: user_id.to_string(),
            file_context: Some(file_context),
            ..Default::default()
        };

        if let Err(e) = store::save_draft(&record, &[], "", "") {
            display::print_error("vscode", &format!("Failed to save empty-window draft: {e}"));
            continue;
        }
        dedup.mark(session_id, &hash);
        new_count += 1;
    }

    if new_count > 0 {
        let last = requests
            .last()
            .map(extract_empty_window_prompt)
            .unwrap_or_default();
        display::print_drafted(&display::DraftDisplayOptions {
            project_name: "(no workspace)",
            prompt_text: &last,
            exchange_count: new_count as u64,
            ..Default::default()
        });
    }
}

fn extract_empty_window_prompt(req: &EmptyWindowReq) -> String {
    req.message
        .parts
        .iter()
        .filter(|p| !p.text.is_empty())
        .map(|p| p.text.clone())
        .collect::<Vec<_>>()
        .join("\n")
}

fn extract_empty_window_response(req: &EmptyWindowReq) -> String {
    req.response
        .iter()
        .filter(|r| !r.value.is_empty())
        .map(|r| r.value.clone())
        .collect::<Vec<_>>()
        .join("\n")
}
